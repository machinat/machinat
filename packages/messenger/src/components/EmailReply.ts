/* eslint-disable import/prefer-default-export */
import { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makePartSegment, PartSegment } from '@sociably/core/renderer';

const EMAIL_QUICK_REPLY_VALUES = { content_type: 'user_email' };

export function EmailReply(
  node: NativeElement<{}, AnyNativeComponent>,
  path: string
): PartSegment<{}>[] {
  return [makePartSegment(node, path, EMAIL_QUICK_REPLY_VALUES)];
}
