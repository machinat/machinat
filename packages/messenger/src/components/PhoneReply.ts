/* eslint-disable import/prefer-default-export */
import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makePartSegment, PartSegment } from '@sociably/core/renderer';

const PHONE_QUICK_REPLY_VALUES = { content_type: 'user_phone_number' };

export function PhoneReply(
  node: NativeElement<{}, AnyNativeComponent>,
  path: string
): PartSegment<{}>[] {
  return [makePartSegment(node, path, PHONE_QUICK_REPLY_VALUES)];
}
