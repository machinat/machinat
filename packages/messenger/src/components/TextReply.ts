import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makePartSegment, PartSegment } from '@sociably/core/renderer';

/** @category Props */
export type TextReplyProps = {
  /** The text to display on the quick reply button. 20 character limit. */
  title: string;
  /**
   * Custom data that will be sent back to you via the messaging_postbacks
   * webhook event. 1000 character limit.
   */
  payload: string;
  /**
   * URL of image to display on the quick reply button for text quick replies.
   * Image should be a minimum of 24px x 24px. Larger images will be
   * automatically cropped and resized. Required if title is an empty string.
   */
  imageUrl?: string;
};

export function TextReply(
  node: NativeElement<TextReplyProps, AnyNativeComponent>,
  path: string,
): PartSegment<{}>[] {
  const { title, payload, imageUrl } = node.props;
  return [
    makePartSegment(node, path, {
      content_type: 'text',
      title,
      payload,
      image_url: imageUrl,
    }),
  ];
}
