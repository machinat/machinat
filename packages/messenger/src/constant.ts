export const MESSENGER = 'messenger' as const;
export const FACEBOOK = 'facebook' as const;

export const ATTACHMENT_DATA = Symbol.for('attachment_data.messenger.sociably');
export const ATTACHMENT_INFO = Symbol.for('attachment_info.messenger.sociably');

export const ATTACHMENT_ASSET_TAG = Symbol.for(
  'attatchment_asset_tag.messenger.sociably'
);
export const API_PATH = Symbol.for('api_path.messenger.sociably');

export const MESSENGER_MESSAGING_TYPE_RESPONSE = 'RESPONSE';

export const PATH_MESSAGES = 'me/messages';

export const PATH_PASS_THREAD_CONTROL = 'me/pass_thread_control';

export const PATH_REQUEST_THREAD_CONTROL = 'me/request_thread_control';

export const PATH_TAKE_THREAD_CONTROL = 'me/take_thread_control';

export const PATH_MESSAGE_ATTACHMENTS = 'me/message_attachments';

export const PATH_MESSENGER_PROFILE = 'me/messenger_profile';

export const PATH_PERSONAS = 'me/personas';
