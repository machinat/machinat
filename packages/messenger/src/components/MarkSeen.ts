/* eslint-disable import/prefer-default-export */
import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import { PATH_MESSAGES } from '../constant.js';
import type { SenderActionValue } from '../types.js';

const MARK_SEEN_VALUE = {
  type: 'message' as const,
  apiPath: PATH_MESSAGES,
  params: { sender_action: 'mark_seen' as const },
};

export function MarkSeen(
  node: NativeElement<{}, AnyNativeComponent>,
  path: string
): UnitSegment<SenderActionValue>[] {
  return [makeUnitSegment(node, path, MARK_SEEN_VALUE)];
}
