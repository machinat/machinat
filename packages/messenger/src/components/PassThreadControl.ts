import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import { PATH_PASS_THREAD_CONTROL } from '../constant.js';
import type { PassThreadControlValue } from '../types.js';

/** @category Props */
export type PassThreadControlProps = {
  /**
   * The app ID to pass thread control to. Required if the Primary Receiver is
   * passing thread control. If the Secondary Receiver is passing thread
   * control, defaults to the app ID of the Primary Receiver. To pass thread
   * control to the Page inbox, use app ID 263902037430900.
   */
  targetAppId: number;
  /**
   * Metadata passed to the receiving app in the pass_thread_control webhook
   * event.
   */
  metadata?: string;
};

export function PassThreadControl(
  node: NativeElement<PassThreadControlProps, AnyNativeComponent>,
  path: string,
): UnitSegment<PassThreadControlValue>[] {
  const { targetAppId, metadata } = node.props;
  return [
    makeUnitSegment(node, path, {
      type: 'message',
      apiPath: PATH_PASS_THREAD_CONTROL,
      params: { target_app_id: targetAppId, metadata },
    }),
  ];
}
