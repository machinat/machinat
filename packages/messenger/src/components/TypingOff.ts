/* eslint-disable import/prefer-default-export */
import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import { PATH_MESSAGES } from '../constant.js';
import type { SenderActionValue } from '../types.js';

const TYPING_OFF_VALUE = {
  type: 'message' as const,
  apiPath: PATH_MESSAGES,
  params: { sender_action: 'typing_off' as const },
};

export function TypingOff(
  node: NativeElement<{}, AnyNativeComponent>,
  path: string,
): UnitSegment<SenderActionValue>[] {
  return [makeUnitSegment(node, path, TYPING_OFF_VALUE)];
}
