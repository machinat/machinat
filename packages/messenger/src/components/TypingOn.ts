/* eslint-disable import/prefer-default-export */
import type { NativeElement, AnyNativeComponent } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import { PATH_MESSAGES } from '../constant.js';
import type { SenderActionValue } from '../types.js';

const TYPING_ON_VALUE = {
  type: 'message' as const,
  apiPath: PATH_MESSAGES,
  params: { sender_action: 'typing_on' as const },
};

export function TypingOn(
  node: NativeElement<{}, AnyNativeComponent>,
  path: string,
): UnitSegment<SenderActionValue>[] {
  return [makeUnitSegment(node, path, TYPING_ON_VALUE)];
}
