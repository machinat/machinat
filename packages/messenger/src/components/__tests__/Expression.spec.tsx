import Sociably from '@sociably/core';
import { PATH_MESSAGES } from '../../constant.js';
import { TextReply as _TextReply } from '../TextReply.js';
import { PhoneReply as _PhoneReply } from '../PhoneReply.js';
import { EmailReply as _EmailReply } from '../EmailReply.js';
import { Expression as _Expression } from '../Expression.js';
import { PassThreadControl as _PassThreadControl } from '../PassThreadControl.js';
import { TypingOff as _TypingOff } from '../TypingOff.js';
import { TypingOn as _TypingOn } from '../TypingOn.js';
import { MarkSeen as _MarkSeen } from '../MarkSeen.js';
import { renderUnitElement, makeTestComponent } from './utils.js';

const Expression = makeTestComponent(_Expression);
const TextReply = makeTestComponent(_TextReply);
const PhoneReply = makeTestComponent(_PhoneReply);
const EmailReply = makeTestComponent(_EmailReply);
const PassThreadControl = makeTestComponent(_PassThreadControl);
const TypingOff = makeTestComponent(_TypingOff);
const TypingOn = makeTestComponent(_TypingOn);
const MarkSeen = makeTestComponent(_MarkSeen);

const quickReplies = (
  <>
    <TextReply title="foo" payload="bar" />
    <PhoneReply />
    <EmailReply />
  </>
);

const children = (
  <>
    <p>foo</p>
    bar
  </>
);

it.each([
  [
    'Expression with type',
    <Expression messagingType="UPDATE">{children}</Expression>,
  ],
  [
    'Expression with type and tag',
    <Expression messagingType="MESSAGE_TAG" tag="ACCOUNT_UPDATE">
      {children}
    </Expression>,
  ],
  [
    'Expression with notificationType',
    <Expression notificationType="NO_PUSH">{children}</Expression>,
  ],
  [
    'Expression with metadata',
    <Expression metadata="_SOME_METADATA_">{children}</Expression>,
  ],
  [
    'Expression with quickReplies',
    <Expression quickReplies={quickReplies}>{children}</Expression>,
  ],
  [
    'Expression with personaId',
    <Expression personaId="_PERSONA_ID_">{children}</Expression>,
  ],
  [
    'Expression with all props',
    <Expression
      messagingType="MESSAGE_TAG"
      tag="CONFIRMED_EVENT_UPDATE"
      notificationType="SILENT_PUSH"
      metadata="_MY_METADATA_"
      quickReplies={quickReplies}
      personaId="_PERSONA_ID_"
    >
      {children}
    </Expression>,
  ],
  [
    'Expression with break, pause and thunk in children',
    <Expression>
      <p>foo</p>
      bar
      <Sociably.Pause time={2000} />
      baz
      <Sociably.Thunk effect={async () => {}} />
    </Expression>,
  ],
])('%s match snapshot', async (_, element) => {
  await expect(renderUnitElement(element)).resolves.toMatchSnapshot();
});

it('returns null if empty children received', async () => {
  await expect(
    renderUnitElement(<Expression>{undefined}</Expression>),
  ).resolves.toBe(null);
  await expect(
    renderUnitElement(<Expression>{null}</Expression>),
  ).resolves.toBe(null);
});

it('hoist text value into message object', async () => {
  const segments = await renderUnitElement(
    <Expression>
      foo
      <p>bar</p>
      {'baz'}
    </Expression>,
  );

  expect(segments).toMatchInlineSnapshot(`
    [
      {
        "node": "foo",
        "path": "$#Expression.children:0",
        "type": "unit",
        "value": {
          "apiPath": "me/messages",
          "params": {
            "message": {
              "text": "foo",
            },
            "messaging_type": undefined,
            "notification_type": undefined,
            "persona_id": undefined,
            "tag": undefined,
          },
          "type": "message",
        },
      },
      {
        "node": <p>
          bar
        </p>,
        "path": "$#Expression.children:1",
        "type": "unit",
        "value": {
          "apiPath": "me/messages",
          "params": {
            "message": {
              "text": "bar",
            },
            "messaging_type": undefined,
            "notification_type": undefined,
            "persona_id": undefined,
            "tag": undefined,
          },
          "type": "message",
        },
      },
      {
        "node": "baz",
        "path": "$#Expression.children:2",
        "type": "unit",
        "value": {
          "apiPath": "me/messages",
          "params": {
            "message": {
              "metadata": undefined,
              "quick_replies": undefined,
              "text": "baz",
            },
            "messaging_type": undefined,
            "notification_type": undefined,
            "persona_id": undefined,
            "tag": undefined,
          },
          "type": "message",
        },
      },
    ]
  `);
});

it('add attributes to message value', async () => {
  const segments = await renderUnitElement(
    <Expression
      messagingType="MESSAGE_TAG"
      tag="HUMAN_AGENT"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
    >
      foo
      <p>bar</p>
      {'baz'}
    </Expression>,
  );

  segments!.forEach((segment) => {
    expect(segment.value).toEqual({
      type: 'message',
      apiPath: PATH_MESSAGES,
      params: {
        messaging_type: 'MESSAGE_TAG',
        tag: 'HUMAN_AGENT',
        notification_type: 'SILENT_PUSH',
        persona_id: '_PERSONA_ID_',
        message: {
          text: expect.any(String),
        },
      },
    });
  });
});

it('add persona_id to typeing_on/typeing_off sender action', async () => {
  const segments = await renderUnitElement(
    <Expression
      messagingType="MESSAGE_TAG"
      tag="ACCOUNT_UPDATE"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
    >
      foo
      <TypingOn />
      <TypingOff />
      <MarkSeen />
    </Expression>,
  );

  expect(segments?.map(({ value }) => value)).toMatchInlineSnapshot(`
    [
      {
        "apiPath": "me/messages",
        "params": {
          "message": {
            "metadata": undefined,
            "quick_replies": undefined,
            "text": "foo",
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "_PERSONA_ID_",
          "tag": "ACCOUNT_UPDATE",
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "persona_id": "_PERSONA_ID_",
          "sender_action": "typing_on",
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "persona_id": "_PERSONA_ID_",
          "sender_action": "typing_off",
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "sender_action": "mark_seen",
        },
        "type": "message",
      },
    ]
  `);
});

it('adds metadata to last message action', async () => {
  const segments = await renderUnitElement(
    <Expression metadata="_META_">
      foo
      <p>bar</p>
      <TypingOn />
    </Expression>,
  );

  expect(segments?.map(({ value }) => value)).toMatchInlineSnapshot(`
    [
      {
        "apiPath": "me/messages",
        "params": {
          "message": {
            "text": "foo",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "message": {
            "metadata": "_META_",
            "quick_replies": undefined,
            "text": "bar",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "persona_id": undefined,
          "sender_action": "typing_on",
        },
        "type": "message",
      },
    ]
  `);
});

it('adds quickReplies to last message action', async () => {
  const segments = await renderUnitElement(
    <Expression
      quickReplies={
        <>
          <TextReply title="foo" payload="bar" />
          <PhoneReply />
          <EmailReply />
        </>
      }
    >
      <p>foo</p>
      bar
      <TypingOn />
    </Expression>,
  );

  expect(segments?.map(({ value }) => value)).toMatchInlineSnapshot(`
    [
      {
        "apiPath": "me/messages",
        "params": {
          "message": {
            "text": "foo",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "message": {
            "metadata": undefined,
            "quick_replies": [
              {
                "content_type": "text",
                "image_url": undefined,
                "payload": "bar",
                "title": "foo",
              },
              {
                "content_type": "user_phone_number",
              },
              {
                "content_type": "user_email",
              },
            ],
            "text": "bar",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "tag": undefined,
        },
        "type": "message",
      },
      {
        "apiPath": "me/messages",
        "params": {
          "persona_id": undefined,
          "sender_action": "typing_on",
        },
        "type": "message",
      },
    ]
  `);
});

it('do nothing to non-messgae value', async () => {
  const segments = await renderUnitElement(
    <Expression
      messagingType="MESSAGE_TAG"
      tag="CONFIRMED_EVENT_UPDATE"
      notificationType="SILENT_PUSH"
      personaId="_PERSONA_ID_"
      quickReplies={<TextReply title="foo" payload="bar" />}
    >
      <PassThreadControl targetAppId={123} />
      foo
    </Expression>,
  );

  expect(segments?.[0].value).toMatchInlineSnapshot(`
    {
      "apiPath": "me/pass_thread_control",
      "params": {
        "metadata": undefined,
        "target_app_id": 123,
      },
      "type": "message",
    }
  `);
  expect(segments?.[1].value).toMatchInlineSnapshot(`
    {
      "apiPath": "me/messages",
      "params": {
        "message": {
          "metadata": undefined,
          "quick_replies": [
            {
              "content_type": "text",
              "image_url": undefined,
              "payload": "bar",
              "title": "foo",
            },
          ],
          "text": "foo",
        },
        "messaging_type": "MESSAGE_TAG",
        "notification_type": "SILENT_PUSH",
        "persona_id": "_PERSONA_ID_",
        "tag": "CONFIRMED_EVENT_UPDATE",
      },
      "type": "message",
    }
  `);
});
