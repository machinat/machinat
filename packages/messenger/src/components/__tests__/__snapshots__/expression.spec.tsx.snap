// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expression with all props match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": "SILENT_PUSH",
      "persona_id": "_PERSONA_ID_",
      "tag": "CONFIRMED_EVENT_UPDATE",
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": "_MY_METADATA_",
        "quick_replies": Array [
          Object {
            "content_type": "text",
            "image_url": undefined,
            "payload": "bar",
            "title": "foo",
          },
          Object {
            "content_type": "user_phone_number",
          },
          Object {
            "content_type": "user_email",
          },
        ],
        "text": "bar",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": "SILENT_PUSH",
      "persona_id": "_PERSONA_ID_",
      "tag": "CONFIRMED_EVENT_UPDATE",
    },
  },
]
`;

exports[`Expression with break, pause and thunk in children match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children:0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": "bar",
    "path": "$#Expression.children:2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": <Sociably.Pause
      time={2000}
    />,
    "path": "$#Expression.children:3",
    "type": "pause",
    "value": [Function],
  },
  Object {
    "node": "baz",
    "path": "$#Expression.children:4",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "baz",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": <Sociably.Thunk
      effect={[Function]}
    />,
    "path": "$#Expression.children:5",
    "type": "thunk",
    "value": [Function],
  },
]
`;

exports[`Expression with metadata match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": "_SOME_METADATA_",
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;

exports[`Expression with notificationType match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": "NO_PUSH",
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": "NO_PUSH",
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;

exports[`Expression with personaId match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": "_PERSONA_ID_",
      "tag": undefined,
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": "_PERSONA_ID_",
      "tag": undefined,
    },
  },
]
`;

exports[`Expression with quickReplies match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": Array [
          Object {
            "content_type": "text",
            "image_url": undefined,
            "payload": "bar",
            "title": "foo",
          },
          Object {
            "content_type": "user_phone_number",
          },
          Object {
            "content_type": "user_email",
          },
        ],
        "text": "bar",
      },
      "messaging_type": undefined,
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;

exports[`Expression with type and tag match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": "ACCOUNT_UPDATE",
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": "MESSAGE_TAG",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": "ACCOUNT_UPDATE",
    },
  },
]
`;

exports[`Expression with type match snapshot 1`] = `
Array [
  Object {
    "node": "foo",
    "path": "$#Expression.children::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "text": "foo",
      },
      "messaging_type": "UPDATE",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
  Object {
    "node": <bar />,
    "path": "$#Expression.children::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "metadata": undefined,
        "quick_replies": undefined,
        "text": "bar",
      },
      "messaging_type": "UPDATE",
      "notification_type": undefined,
      "persona_id": undefined,
      "tag": undefined,
    },
  },
]
`;
